### ISSUE TITLE
Indexing problem using pretrained word embedding

### ISSUE THREAD

**Issue by RikiConrey** [rstudio/keras#302]
The author of the pretrained word embedding example `[here](https://github.com/rstudio/keras/blob/master/vignettes/examples/pretrained_word_embeddings.R)`. Notes that the example does not work; it under-performs the same example in Python. 

This is because there is an indexing problem with the embedding layer. 

The original code uses the tokenizer word index as the R-style column index (1...n), and that model fails. 

```{r}
prepare_embedding_matrix <- function() {
  embedding_matrix <- matrix(0L, nrow = num_words, ncol = EMBEDDING_DIM)
  for (word in names(word_index)) {
    index <- word_index[[word]]
    if (index >= MAX_NUM_WORDS)
      next
    embedding_vector <- embeddings_index[[word]]
    if (!is.null(embedding_vector)) {
      # words not found in embedding index will be all-zeros.
      embedding_matrix[index,] <- embedding_vector
    }
  }
  embedding_matrix
}
```

Epoch 10/10
15998/15998 [==============================] - 107s 7ms/step - loss: 1.0706 - acc: 0.6563 - val_loss: 2.0725 - val_acc: 0.4036

Adding 1 row to the embedding weights and leaving the top row (0 in Python, 1 in R) all 0's substantially improves the model performance. 

```{r}
prepare_embedding_matrix <- function() {
  embedding_matrix <- matrix(0L, nrow = num_words+1, ncol = EMBEDDING_DIM)
  for (word in names(word_index)) {
    index <- word_index[[word]]
    if (index >= MAX_NUM_WORDS)
      next
    embedding_vector <- embeddings_index[[word]]
    if (!is.null(embedding_vector)) {
      # words not found in embedding index will be all-zeros.
      embedding_matrix[index+1,] <- embedding_vector
    }
  }
  embedding_matrix
}
```

`embedding_layer <- layer_embedding(
  input_dim = num_words+1,
  output_dim = EMBEDDING_DIM,
  weights = list(embedding_matrix),
  input_length = MAX_SEQUENCE_LENGTH,
  trainable = FALSE
)
`

Epoch 10/10
15998/15998 [==============================] - 105s 7ms/step - loss: 0.3726 - acc: 0.8727 - val_loss: 1.0579 - val_acc: 0.6957

I'm not 100% sure why this is happening (it seems like an indexing translation problem between the two languages), but I am pretty sure this is the solution.

**Comment by jjallaire** [comment]
Thanks for pointing this out! I discovered this as well when working on the chapter on pretrained embeddings in the DL w/ R book. I'll try to update the example with this fix soon.

**Comment by tmerrittsmith** [comment]
Thanks @RikiConrey I've been pondering this for some time now!
